Objective-C/CBLDatabase.mm

```diff
@@ -950,102 +952,105 @@ - (BOOL) saveDocument: (CBLDocument*)document
-    CBL_LOCK(self) {
-        C4Transaction t(_c4db);
-        if (!t.begin())
-            return convertError(t.error(), outError);
-        
-        // Read local document:
-        CBLDocument* localDoc = [[CBLDocument alloc] initWithDatabase: self
-                                                           documentID: docID
-                                                       includeDeleted: YES
-                                                                error: outError];
-        if (!localDoc)
-            return false;
+- (bool) resolveConflictInDocument: (NSString*)docID
+              withConflictResolver: (id<CBLConflictResolver>)conflictResolver
+                             error: (NSError**)outError {
+    while (true) {
+        CBLDocument* localDoc;
+        CBLDocument* remoteDoc;
+        CBL_LOCK(self) {
+            C4Transaction t(_c4db);
+            if (!t.begin())
+                return convertError(t.error(), outError);
+            
+            // Read local document:
+            localDoc = [[CBLDocument alloc] initWithDatabase: self documentID: docID
+                                              includeDeleted: YES error: outError];
+            if (!localDoc)
+                return false;
+            
+            // Read the conflicting remote revision:
+            remoteDoc = [[CBLDocument alloc] initWithDatabase: self documentID: docID
+                                               includeDeleted: YES error: outError];
+            if (!remoteDoc || ![remoteDoc selectConflictingRevision])
+                return false;
+            
+            if (!t.commit())
+                return convertError(t.error(), outError);
+        }
         
-        // Read the conflicting remote revision:
-        CBLDocument* remoteDoc = [[CBLDocument alloc] initWithDatabase: self
-                                                            documentID: docID
-                                                        includeDeleted: YES
-                                                                 error: outError];
-        if (!remoteDoc || ![remoteDoc selectConflictingRevision])
-            return false;
+        conflictResolver = conflictResolver ?: [CBLConflictResolution default];
+        CBLConflict* conflict = [[CBLConflict alloc] initWithLocalDocument: localDoc
+                                                            remoteDocument: remoteDoc];
         
         // Resolve conflict:
         CBLLogInfo(Sync, @"Resolving doc '%@' (mine=%@ and theirs=%@)",
                    docID, localDoc.revID, remoteDoc.revID);
+        CBLDocument* resolvedDoc = [conflictResolver resolve: conflict];
         
-        CBLDocument* resolvedDoc = [self resolveConflictBetweenLocalDoc: localDoc
-                                                           andRemoteDoc: remoteDoc];
+        if (resolvedDoc == nil) {
+            resolvedDoc = remoteDoc.isDeleted ? remoteDoc : localDoc;
```

Comment:
It's possible that both remoteDoc and localDoc are not deleted doc. There might not be a real use case but the API allows it. I think you could just pass null to saveResolvedDocument() then saveResolvedDocument() will determine it as a deleted doc.

Reply:
done